lib/plugins/threaded-signing/plugin.c

```diff
@@ -254,30 +283,32 @@ threaded_openssl_get_signature(sv_threaded_plugin_t *self,
   SignedVideoReturnCode status = SV_OK;
 
   g_mutex_lock(&self->mutex);
-  if (self->plugin_state == THREADED_SIGNING_HAS_SIGNATURE && self->signature_info) {
-    if (self->signature_info->signature_size > max_signature_size) {
+  if (self->output_buffer[0].signing_error) {
+    *written_signature_size = 0;
+    // Propagate SV_EXTERNAL_ERROR when signing failed.
+    status = SV_EXTERNAL_ERROR;
+  } else if (self->output_buffer_idx > 0) {
+    if (self->output_buffer[0].size > max_signature_size) {
       // If there is no room to copy the signature, report zero size.
       *written_signature_size = 0;
     } else {
-      memcpy(signature, self->signature_info->signature, self->signature_info->signature_size);
-      *written_signature_size = self->signature_info->signature_size;
+      // Get the oldest signature
+      memcpy(signature, self->output_buffer[0].signature, self->output_buffer[0].size);
+      *written_signature_size = self->output_buffer[0].size;
+      // Change state and mark as copied.
+      has_copied_signature = true;
     }
-    // Change state and mark as copied.
-    self->plugin_state = THREADED_SIGNING_WAITS_FOR_HASH_TO_SIGN;
-    has_copied_signature = true;
-  } else if (self->plugin_state == THREADED_SIGNING_ERROR) {
-    *written_signature_size = 0;
-    // Change state and mark as copied.
-    self->plugin_state = THREADED_SIGNING_WAITS_FOR_HASH_TO_SIGN;
-    has_copied_signature = true;
-    // Propagate SV_EXTERNAL_ERROR when signing failed.
-    status = SV_EXTERNAL_ERROR;
-  } else if (self->plugin_state == THREADED_SIGNING_WAITS_FOR_HASH_TO_SIGN &&
-      self->nbr_of_unsigned_hashes > 0) {
-    // There are unsigned hashes in the pipe. Report them with zero size, since no signature exists.
-    *written_signature_size = 0;
-    self->nbr_of_unsigned_hashes--;
-    has_copied_signature = true;
+  }
+  if (self->output_buffer_idx > 0) {
+    // Move buffer
+    output_data_t tmp = self->output_buffer[0];
+    int i = 0;
+    while (self->output_buffer[i].signature != NULL) {
```

Comment:
Like for the ```input_buffer```, you will read and write outside memory if the buffer is full of allocated pointers.